feat: Add Client -> Connection layer

devjgm opened the pull request:
Fixes #305 
Related to #196 

This PR adds an abstract `Connection` interface that gives callers the ability to mock Client's behavior. This functionality is need for Alpha (#196). `Client` now holds a shared ptr to a `Connection` instance. And a `Connection` holds a shared pointer to a `SpannerStub`. See the more detailed design rationale in #305.

The main changes in this PR are that most of the *implementation* from `client.cc` moved to `internal/connection_impl.cc`. And the unit test stuff as well. This mostly affected `Commit`.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/googleapis/google-cloud-cpp-spanner/320)
<!-- Reviewable:end -->

devbww commented on google/cloud/spanner/internal/connection_impl.h:
```diff
@@ -0,0 +1,77 @@
+  StatusOr<CommitResult> Commit(google::spanner::v1::TransactionSelector& s,
+                                std::vector<Mutation> mutations);
+  std::string database_;
+  std::shared_ptr<internal::SpannerStub> stub_;
```
Why do we choose to make this shared?  gRPC gave it to us as a `unique_ptr`, which we converted into a `shared_ptr`, but (as far as I can see) its refcount remains at 1.

devbww replied:
> Don't know. I'm just copying existing code. Do you mind if we follow up on this change in a separate PR?

I'd suggest that it should be unique until it needs to be otherwise, but a separate PR is good.

devbww commented on google/cloud/spanner/internal/connection_impl.h:
```diff
@@ -0,0 +1,77 @@
+                          std::shared_ptr<internal::SpannerStub> stub)
+      : database_(std::move(database)), stub_(std::move(stub)) {}
+
+  StatusOr<CommitResult> Commit(CommitParams) override;
```
Should we use `final`?

devbww replied:
> Don't know. Maybe? I don't have a sense of when to apply `final`. Suggestion?

My only suggestions would be to forget about it, or apply it to every override (unless that contradicts some explicit need).  I'm happy with the former for now.

devbww commented on google/cloud/spanner/connection.h:
```diff
@@ -0,0 +1,67 @@
+   * commonly, the cause is conflicts with concurrent transactions. However, it
+   * can also happen for a variety of other reasons. If `Commit` returns
+   * `ABORTED`, the caller should re-attempt the transaction from the beginning,
+   * re-using the same session.
```
Is the "same session" part important or even necessary?

devbww replied:
> It is important. Re-using the same session is how Spanner guarantees that the every Transaction will eventually "win" and be serviced.

Ack.  We mused about including that information, or an link, here.

mr-salty commented:
___
*[google/cloud/spanner/connection.h, line 46 at r5](https://reviewable.io/reviews/googleapis/google-cloud-cpp-spanner/320#-LlinKGa81ew-lMn_UWV:-LlinKGa81ew-lMn_UWW:bel7dfl) ([raw file](https://github.com/googleapis/google-cloud-cpp-spanner/blob/144c45f6c8bbb7df86c36e662ec132ae1ed18234/google/cloud/spanner/connection.h#L46)):*
<details><summary><i alt="summary">Quoted 8 lines of code…</i></summary>

> ```
>   /**
>    * Commits a transaction.
>    *
>    * The commit might return an `ABORTED` error. This can occur at any time;
>    * commonly, the cause is conflicts with concurrent transactions. However, it
>    * can also happen for a variety of other reasons. If `Commit` returns
>    * `ABORTED`, the caller should re-attempt the transaction from the beginning,
>    * re-using the same session.
> ```
</details>


I don't think we should duplicate the documentation from `Client` here.


<!-- Sent from Reviewable.io -->

mr-salty commented:
___
*[google/cloud/spanner/internal/connection_impl.h, line 64 at r3](https://reviewable.io/reviews/googleapis/google-cloud-cpp-spanner/320#-LlggtS9xHIoDLQEshUx-r3-64:-LlilI9k6EaGfNSlEOrc:by5ulo8) ([raw file](https://github.com/googleapis/google-cloud-cpp-spanner/blob/a6ba272a5ca7e5e3096db25790dcec1884aaf62b/google/cloud/spanner/internal/connection_impl.h#L64)):*
<details><summary><i>Previously, devbww (Bradley White) wrote…</i></summary><blockquote>

> Don't know. I'm just copying existing code. Do you mind if we follow up on this change in a separate PR?

I'd suggest that it should be unique until it needs to be otherwise, but a separate PR is good.
</blockquote></details>

- This is not the gRPC stub; that is held as a `unique_ptr` inside `SpannerStub`.
- The design doc I was working from specified that `Stub` should be a `shared_ptr`. In general I didn't attempt to re-evaluate those decisions when I wrote the code; we can certainly revisit, but we should understand the rationale (perhaps it is no longer be applicable).


<!-- Sent from Reviewable.io -->

mr-salty commented:
Sorry for not commenting while the PR was open, I didn't realize it was on the verge of being submitted.





<!-- Sent from Reviewable.io -->

devjgm commented:
No problem. I'll do a follow-up PR addressing your comments.

On Wed, Aug 7, 2019, 9:22 PM Todd Derr <notifications@github.com> wrote:

> Sorry for not commenting while the PR was open, I didn't realize it was on
> the verge of being submitted.
>
> —
> You are receiving this because you modified the open/close state.
> Reply to this email directly, view it on GitHub
> <https://github.com/googleapis/google-cloud-cpp-spanner/pull/320?email_source=notifications&email_token=ACICRW32YFWSVCN6ZL46CYDQDNYOPA5CNFSM4IKBVTB2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOD32ERLY#issuecomment-519325871>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ACICRWYCLOC5XNXP7FCEUX3QDNYOPANCNFSM4IKBVTBQ>
> .
>
